### PR TITLE
Warrior Sim UI

### PR DESCRIPTION
**Pull request by TheGroxEmpire**
Finished:
- [x] Warrior background / UI style
- [x] Appropriate EP Weights & stats
- [x] Buffs & debuffs that are available
- [x] Consumes that are available
- [x] Fury, Arms Slam, Arms DW talents
- [x] P1 presets bis for all three specs, minus non available gems
- [x] Rotation options for all three specs

Features that's missing:
- [ ] Complete gems on preset: Smooth gems (Yellow crit gem) and Jagged gems (Green crit gem)
- [ ] Complete buffs: Windfury totem, Strength of Earth totem, and Unleashed Rage
- [ ] Complete debuffs: Improved Hunter's Mark
- [ ] Complete consumes: Mighty Rage potion and Elixir of the Mongoose

### CODE REVIEW

File: proto/warrior.proto [hunk 1]
```diff
@@ -57,12 +57,62 @@ message WarriorTalents {
 
 message Warrior {
     message Rotation {
+		enum Type {
+			ArmsSlam = 0;
+			ArmsDW = 1;
+			Fury = 2;
+			General = 3;
```
Comment (jimmyt857) [inline proto/warrior.proto:64]: I'm assuming that the 'General' category is fields shared by all rotations? If so you dont need an enum and separate GeneralRotation message to hold them, just put them alongside the other fields in Rotation.

File: proto/warrior.proto [hunk 2]
```diff
@@ -57,12 +57,62 @@ message WarriorTalents {
...
+		}
+
+		Type type = 1;
+
+		message ArmsSlamRotation {
+
```
Comment (jimmyt857) [inline proto/warrior.proto:70]: Nit: Remove blank line.

File: proto/warrior.proto [hunk 3]
```diff
@@ -57,12 +57,62 @@ message WarriorTalents {
...
+		message FuryRotation {
+			enum PrimaryInstant {
+				Bloodthirst = 0;
+				Whirlwind = 1;
+			}
+			PrimaryInstant primary_instant = 1;
+
+			bool use_bt_exec = 2;
+			double rampage_cd_tresh = 10; // only when using rampage
```
Comment (jimmyt857) [inline proto/warrior.proto:91]: Might be a misspelling, should that be 'rampage_cd_thresh'?

If so, if you don't mind lets use 'rampage_cd_threshold' instead. Proto messages are the main interface so variables should be named to be immediately obvious to anyone (even non-warriors  ;)   )

File: proto/warrior.proto [hunk 4]
```diff
@@ -57,12 +57,62 @@ message WarriorTalents {
...
+		FuryRotation fury = 4;
+
+		message GeneralRotation {
+			bool use_ww_exec = 1;
+			bool use_hs_exec = 2;
+			double hs_rage_thresh = 3;
+			bool use_overpower = 4;
+			double overpower_rage_thresh = 5;
+			bool use_hamstring = 6;
+			double hamstring_rage_thresh = 7;
```
Comment (jimmyt857) [inline proto/warrior.proto:103]: These are the variables I referred to before, you can remove the 'GeneralRotation' message and just have these within the Rotation message.

Also, I'm seeing the naming pattern '_thresh' and 'use_xx_exec' a lot, could we make them '_threshold' and 'use_xx_during_execute' instead? I know its a little verbose but it helps a lot if a non-warrior has to read your code.

File: ui/warrior/inputs.ts [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+		//                       FURY INPUTS
+		// ********************************************************
+		{
+			type: 'enum' as const,
+			getModObject: (simUI: IndividualSimUI<any>) => simUI.player,
+			config: {
+				extraCssClasses: [
+					'rotation-type-enum-picker',
+				],
+				label: 'Primary Instant',
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:99]: A tooltip here might be a good idea, just to clarify what 'Primary Instant' does exactly. I assume something like, 'Highest priority ability to use. The other one will still be used when available and the primary is on CD.'

File: ui/warrior/presets.ts [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+		overpowerRageThresh: 25,
+		useHamstring: false,
+		hamstringRageThresh: 70,
+	}),
+});
+
+export const P1_FURY_PRESET = {
+	name: 'P1 Fury Preset',
+	tooltip: Tooltips.BASIC_BIS_DISCLAIMER,
+	enableWhen: (player: Player<Spec.SpecWarrior>) => player.getRotation().type == RotationType.Fury,
```
Comment (jimmyt857) [inline ui/warrior/presets.ts:112]: I'm not going to read through all the item IDs here, but just want to clarify that you dont NEED to have separate presets for each spec. You can change or remove the enableWhen condition if some or all of the specs use the same presets.

If you actually want separate presets for each spec then this is fine.

File: ui/warrior/sim.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+			raidBuffInputs: [
+				IconInputs.GiftOfTheWild,
+			],
+			partyBuffInputs: [
+				IconInputs.LeaderOfThePack,
+				IconInputs.DrumsOfBattleBuff,
+				IconInputs.Bloodlust,
+				// IconInputs.WindfuryTotem,
+				// IconInputs.StrengthOfEarthTotem,
+				// IconInputs.UnleashedRage,
```
Comment (jimmyt857) [inline ui/warrior/sim.ts:124]: I'll try to get these (and the consume ones below) added sometime this week!

File: package.json [hunk 8]
```diff
@@ -11,6 +11,7 @@
         "@protobuf-ts/plugin": "^2.0.4",
         "@protobuf-ts/runtime": "^2.0.4",
         "http-server": "13.0.1",
+        "npm": "^8.3.1",
         "sass": "^1.39.0",
         "typescript": "4.3.5"
     }
```
Comment (jimmyt857) [inline package.json:17]: It looks like you have some local workspace changes here (and in package-lock.json) that shouldn't be committed.

**Review by jimmyt857 — CHANGES_REQUESTED** [review]

File: package.json (hunk 8 above)
Comment (TheGroxEmpire) [reply package.json]: Those files have been reverted back now.

File: ui/warrior/presets.ts (hunk 6 above)
Comment (TheGroxEmpire) [reply ui/warrior/presets.ts:112]: I do want them to have different presets. These specs have different bis sets in while the differences between them are minor.

File: ui/warrior/inputs.ts (hunk 5 above)
Comment (TheGroxEmpire) [reply ui/warrior/inputs.ts:99]: Added tooltip on 2dc332d.

File: ui/warrior/sim.ts (hunk 7 above)
Comment (TheGroxEmpire) [reply ui/warrior/sim.ts:124]: It'd be rad if you could also include the missing gems such as Smooth Dawnstone and Jagged Talasite.

File: proto/warrior.proto (hunk 1 above)
Comment (TheGroxEmpire) [reply proto/warrior.proto]: Done in a21327c1.

File: proto/warrior.proto (hunk 2 above)
Comment (TheGroxEmpire) [reply proto/warrior.proto]: Done in a21327c1.

File: proto/warrior.proto (hunk 3 above)
Comment (TheGroxEmpire) [reply proto/warrior.proto]: Renamed 'rampage_cd_tresh' to 'rampage_cd_threshold' in a21327c1.

File: proto/warrior.proto (hunk 4 above)
Comment (TheGroxEmpire) [reply proto/warrior.proto]: Done in a21327c1.

File: ui/warrior/inputs.ts [hunk 9]
```diff
@@ -258,16 +259,13 @@ export const WarriorRotationConfig = {
 		cssClass: 'ww-exec-picker',
 		getModObject: (simUI: IndividualSimUI<any>) => simUI.player,
 		config: {
-			label: 'WW on Execute Phase',
-			labelTooltip: 'Use Whirlwind on Execute Phase.',
+			label: 'WW during Execute Phase',
+			labelTooltip: 'Use Whirlwind during Execute Phase.',
 			changedEvent: (player: Player<Spec.SpecWarrior>) => player.rotationChangeEmitter,
-			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.useWwExec || false,
+			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().useWwDuringExecute || true,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:265]: I don't think you need the '|| true' now that this isn't inside of 'general'.

File: ui/warrior/inputs.ts [hunk 10]
```diff
@@ -277,16 +275,13 @@ export const WarriorRotationConfig = {
 		cssClass: 'hs-exec-picker',
 		getModObject: (simUI: IndividualSimUI<any>) => simUI.player,
 		config: {
-			label: 'HS on Execute Phase',
-			labelTooltip: 'Use Heroic Strike on Execute Phase.',
+			label: 'HS during Execute Phase',
+			labelTooltip: 'Use Heroic Strike during Execute Phase.',
 			changedEvent: (player: Player<Spec.SpecWarrior>) => player.rotationChangeEmitter,
-			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.useHsExec || false,
+			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().useHsDuringExecute || true,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:281]: Same as above

File: ui/warrior/inputs.ts [hunk 11]
```diff
@@ -299,13 +294,10 @@ export const WarriorRotationConfig = {
 			label: 'HS rage threshold',
 			labelTooltip: 'Queue HS when rage is above:',
 			changedEvent: (player: Player<Spec.SpecWarrior>) => player.rotationChangeEmitter,
-			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.hsRageThresh || 60,
+			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().hsRageThreshold || 70,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:297]: When any of these values is missing, its important to return a zero/false/empty value or there will be a desync between the visuals and the internal representation. Actual default values are handled by your Presets.DefaultXXX fields.

File: ui/warrior/inputs.ts [hunk 12]
```diff
@@ -318,13 +310,10 @@ export const WarriorRotationConfig = {
 			label: 'Use Overpower',
 			labelTooltip: 'Use Overpower when it is possible.',
 			changedEvent: (player: Player<Spec.SpecWarrior>) => player.rotationChangeEmitter,
-			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.useOverpower || false,
+			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().useOverpower || false,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:313]: Same as above.

File: ui/warrior/inputs.ts [hunk 13]
```diff
@@ -337,16 +326,13 @@ export const WarriorRotationConfig = {
 			label: 'Overpower rage threshold',
 			labelTooltip: 'Use Overpower when rage is below a point.',
 			changedEvent: (player: Player<Spec.SpecWarrior>) => player.rotationChangeEmitter,
-			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.overpowerRageThresh || 0,
+			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().overpowerRageThreshold || 25,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:329]: Same as above, just remove the '|| 25'.

File: ui/warrior/inputs.ts [hunk 14]
```diff
@@ -337,16 +326,13 @@ export const WarriorRotationConfig = {
...
 				const newRotation = player.getRotation();
-				if (!newRotation.general) {
-					newRotation.general = GeneralRotation.clone(Presets.DefaultGeneralRotation.general!);
-				}
-				newRotation.general.overpowerRageThresh = newValue;
+				newRotation.overpowerRageThreshold = newValue;
 				player.setRotation(eventID, newRotation);
 			},
-			showWhen: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.useOverpower == true,
+			showWhen: (player: Player<Spec.SpecWarrior>) => player.getRotation().useOverpower == true,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:335]: Just '...useOverpower' is fine, don't need the == true.

File: ui/warrior/inputs.ts [hunk 15]
```diff
@@ -357,13 +343,10 @@ export const WarriorRotationConfig = {
 			label: 'Use Hamstring',
 			labelTooltip: 'Use Hamstring on free global.',
 			changedEvent: (player: Player<Spec.SpecWarrior>) => player.rotationChangeEmitter,
-			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.useHamstring || false,
+			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().useHamstring || false,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:346]: Don't need the || false.

File: ui/warrior/inputs.ts [hunk 16]
```diff
@@ -376,16 +359,13 @@ export const WarriorRotationConfig = {
 			label: 'Hamstring rage threshold',
 			labelTooltip: 'Use Hamstring when rage is above a ',
 			changedEvent: (player: Player<Spec.SpecWarrior>) => player.rotationChangeEmitter,
-			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.hamstringRageThresh || 0,
+			getValue: (player: Player<Spec.SpecWarrior>) => player.getRotation().hamstringRageThreshold || 70,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:362]: Remove the || 70.

File: ui/warrior/inputs.ts [hunk 17]
```diff
@@ -376,16 +359,13 @@ export const WarriorRotationConfig = {
...
 				const newRotation = player.getRotation();
-				if (!newRotation.general) {
-					newRotation.general = GeneralRotation.clone(Presets.DefaultGeneralRotation.general!);
-				}
-				newRotation.general.hamstringRageThresh = newValue;
+				newRotation.hamstringRageThreshold = newValue;
 				player.setRotation(eventID, newRotation);
 			},
-			showWhen: (player: Player<Spec.SpecWarrior>) => player.getRotation().general?.useHamstring == true,
+			showWhen: (player: Player<Spec.SpecWarrior>) => player.getRotation().useHamstring == true,
```
Comment (jimmyt857) [inline ui/warrior/inputs.ts:368]: Remove the == true

**Review by jimmyt857 — CHANGES_REQUESTED** [review]

**Comment by TheGroxEmpire** [comment]
All suggestions are applied in eee35046.

**Review by jimmyt857 — APPROVED** [review]

**Pull request merged**